File: xephang.cpp
#include<bits/stdc++.h>
using namespace std;
int main(){
    ios_base::sync_with_stdio(false);cin.tie(0);
    int  n;
    cin>>n;
    stack<pair<long long, long long>> a;
    a.push({LLONG_MAX, 0});
    long long res=0, i=0;
    while (i<n)
    {
        int x;
        cin >>x;
        while (a.top().first<x)
        {
            res+=a.top().second;
            a.pop();
        }
        if(a.top().first==x){
            a.push({x, a.top().second+1});
            res++;
        }else{
            if(a.size()>1) res++;
            a.push({x,1});
        }
        i++;
    }
    cout<<res;
}

File: Xephang.py
res = 0
stack = []

n = int(input())
heights = list(map(int, input().split()))

for height in heights:
    while stack and stack[-1][0] < height:
        res += stack.pop()[1]
    
    if stack and stack[-1][0] == height:
        res += stack[-1][1] + (len(stack) > 1)
        stack[-1][1] += 1
    else:
        res += len(stack) > 0
        stack.append([height, 1])

print(res)
